#  Load data and labels

In [1]:
import lightgbm as lgb

In [2]:
import pandas as pd

#### load training data

In [3]:
train_data = pd.read_csv('orange_small_train.data', sep = '\t')





Var1  Var2  Var3  Var4  Var5    Var6  Var7  Var8  Var9  Var10  ...  \
0   NaN   NaN   NaN   NaN   NaN  1526.0   7.0   NaN   NaN    NaN  ...   
1   NaN   NaN   NaN   NaN   NaN   525.0   0.0   NaN   NaN    NaN  ...   
2   NaN   NaN   NaN   NaN   NaN  5236.0   7.0   NaN   NaN    NaN  ...   
3   NaN   NaN   NaN   NaN   NaN     NaN   0.0   NaN   NaN    NaN  ...   
4   NaN   NaN   NaN   NaN   NaN  1029.0   7.0   NaN   NaN    NaN  ...   
5   NaN   NaN   NaN   NaN   NaN   658.0   7.0   NaN   NaN    NaN  ...   
6   NaN   NaN   NaN   NaN   NaN  1680.0   7.0   NaN   NaN    NaN  ...   
7   NaN   NaN   NaN   NaN   NaN    77.0   0.0   NaN   NaN    NaN  ...   
8   NaN   NaN   NaN   NaN   NaN  1176.0   7.0   NaN   NaN    NaN  ...   
9   NaN   NaN   NaN   NaN   NaN  1141.0   7.0   NaN   NaN    NaN  ...   

    Var221   Var222      Var223  Var224  Var225  Var226   Var227  \
0     oslk  fXVEsaq  jySVZNlOJy     NaN     NaN    xb3V     RAYp   
1     oslk  2Kb5FSF  LM8l689qOp     NaN     NaN    fKCe     RAYp   
2  Al6ZaUT  NKv4yOc  jySVZNlOJy     NaN    kG3k    Qu4f  02N6s8f   
3     oslk  CE7uk3u  LM8l689qOp     NaN     NaN    FSa2     RAYp   
4     oslk  1J2cvxe  LM8l689qOp     NaN    kG3k    FSa2     RAYp   
5     zCkv  QqVuch3  LM8l689qOp     NaN     NaN    Qcbd  02N6s8f   
6     oslk  XlgxB9z  LM8l689qOp     NaN    kG3k    FSa2     RAYp   
7     oslk  R2LdzOv         NaN     NaN     NaN    FSa2     RAYp   
8     zCkv  K2SqEo9  jySVZNlOJy     NaN    kG3k    PM2D     6fzt   
9     oslk  EPqQcw6  LM8l689qOp     NaN    kG3k    FSa2     RAYp   

                   Var228  Var229  Var230  
0           F2FyR07IdsN7I     NaN     NaN  
1           F2FyR07IdsN7I     NaN     NaN  
2           ib5G6X1eUxUn6    am7c     NaN  
3           F2FyR07IdsN7I     NaN     NaN  
4           F2FyR07IdsN7I    mj86     NaN  
5                 Zy3gnGM    am7c     NaN  
6                 55YFVY9    am7c     NaN  
7           F2FyR07IdsN7I     NaN     NaN  
8  am14IcfM7tWLrUmRT52KtA    am7c     NaN  
9                 55YFVY9     NaN     NaN  

[10 rows x 230 columns]

# Data pre-processing

#### Data analysis

In [ ]:
train.describe()

In [ ]:
train = train.fillna(train.mean())

In [ ]:
train.describe()

In [ ]:
train.shape[1]

190~230 columns are strings. Remove them so that NN can hanle.

In [ ]:
droplist = []
for i in range(train.shape[1]-40,train.shape[1]+1):
    value = 0
    value = 'Var{}'.format(i)
    droplist.append(value)

In [ ]:
train = train.drop(droplist,axis=1)
train = train.fillna(0)
train.head(10)

In [ ]:
import matplotlib.pyplot as plt

In [ ]:
%matplotlib inline

In [ ]:
labels = pd.concat([appetency, churn, upselling],axis=1)

In [ ]:
labels.columns = ['appetency', 'churn', 'upselling']

In [ ]:
fig = plt.figure(figsize=(30, 10))
plt.subplot(131)
ax0 = appetency.hist(bins=4)
ax0.set_ylabel('{}'.format('Appetency'))

plt.subplot(132)
ax1= churn.hist(bins=4)
ax1.set_ylabel('{}'.format('Churn'))               

plt.subplot(133)
ax2 = upselling.hist(bins=4)
ax2.set_ylabel('{}'.format('Upselling'))

plt.legend()

In [4]:
train_data.loc[:,'Var228'].head(10)

0             F2FyR07IdsN7I
1             F2FyR07IdsN7I
2             ib5G6X1eUxUn6
3             F2FyR07IdsN7I
4             F2FyR07IdsN7I
5                   Zy3gnGM
6                   55YFVY9
7             F2FyR07IdsN7I
8    am14IcfM7tWLrUmRT52KtA
9                   55YFVY9
Name: Var228, dtype: object

In [5]:
train_data.loc[:,'Var191':'Var230']

Var191      Var192                              Var193 Var194  \
0        NaN  bZkvyxLkBI                                RO12    NaN   
1        NaN  CEat0G8rTN                                RO12    NaN   
2        NaN  eOQt0GoOh3                             AERks4l   SEuy   
3        NaN  jg69tYsGvO                                RO12    NaN   
4        NaN  IXSgUHShse                                RO12   SEuy   
5        NaN  m9SrEy7Rm6                             2Knk1KF    NaN   
6        NaN  Qu0qrQKzJV                             2Knk1KF   lvza   
7        NaN  eOQgUHShse                                RO12    NaN   
8        NaN  4e7Wq69R_D                                RO12   SEuy   
9        NaN  J9Vr4RQZiT                             2Knk1KF   SEuy   
10       NaN  UCDvyx8zhV                          g62hiBSaKg   SEuy   
11       NaN  crIgUHSK8h                                RO12   SEuy   
12       NaN  O8Zr4Rdc_a  onTuEhrJJQy_H3IHkZku5AFczhYGqxJ890    NaN   
13       NaN  cFHt0GoOh3                                RO12   SEuy   
14       NaN  2jiWq69cTy                                RO12    NaN   
15       NaN  _v2TmBftjz                             2Knk1KF    NaN   
16       NaN  FoxgUHSK8h  onTuEhrJJQy_H3IHkZku5AFczhYGqxJ890   SEuy   
17       NaN  FAJgUH7ejg                             AERks4l    NaN   
18       NaN  TfiWq69R_D                                RO12   lvza   
19       NaN  EsYrjIX0Db                                RO12    NaN   
20       NaN  75lq9ayfAo                                RO12    NaN   
21       NaN  TfiWq69R_D                                RO12    NaN   
22       NaN  vAsgUHXM47                                RO12    NaN   
23       NaN  DQlgUH7IEC                                RO12    NaN   
24       NaN  NESt0G8EIb                                RO12   SEuy   
25       NaN  Mtpt0G8EIb                                RO12    NaN   
26      r__I  639TmBQxul                                RO12    NaN   
27       NaN  xOXr4RXktW                                RO12   SEuy   
28       NaN  7Cwt0Gdc_a                             2Knk1KF    NaN   
29       NaN  Yddq9ayE15                                RO12    NaN   
...      ...         ...                                 ...    ...   
24970    NaN  dTGmfo8zhV                                RO12    NaN   
24971    NaN  51rBcmz45j                             2Knk1KF    NaN   
24972    NaN  m9SvyxLkBI                          g62hiBSaKg    NaN   
24973    NaN  NESWq69R_D                          g62hiBSaKg   SEuy   
24974    NaN  1fZWq6e2Rv                             2Knk1KF   SEuy   
24975    NaN  Yddq9ayE15                             2Knk1KF    NaN   
24976    NaN  Foxt0Go8Lb                             2Knk1KF   SEuy   
24977    NaN  1fZWq6e2Rv                                RO12    NaN   
24978    NaN  xearjIX0Db                                RO12    NaN   
24979    NaN  52Iq9ayE15                                RO12    NaN   
24980    NaN  64nWq69R_D                                RO12    NaN   
24981    NaN  DHeq9ayfAo                                RO12   SEuy   
24982    NaN  AUUt0GoOh3                                RO12    NaN   
24983    NaN  JMGq9aXt5a                                RO12   SEuy   
24984    NaN  DHeq9ayfAo                                RO12    NaN   
24985    NaN  vAsvyx7ejg                                RO12    NaN   
24986    NaN  a2iBLJED1x                                RO12   SEuy   
24987    NaN  V8lt0G8rTN                                RO12    NaN   
24988    NaN  NQ5vyxLkBI                                RO12    NaN   
24989    NaN  py5gUH7ejg                             2Knk1KF    NaN   
24990    NaN  tr0rEyXktW                             2Knk1KF   SEuy   
24991    NaN  Iskvyx875g                                RO12    NaN   
24992    NaN  _v2TmBftjz                                RO12   SEuy   
24993    NaN  2jiWq69cTy                             2Knk1KF    NaN   
24994    NaN  52Iq9ayE15      

In [6]:
train_data.dtypes

Var1      float64
Var2      float64
Var3      float64
Var4      float64
Var5      float64
Var6      float64
Var7      float64
Var8      float64
Var9      float64
Var10     float64
Var11     float64
Var12     float64
Var13     float64
Var14     float64
Var15     float64
Var16     float64
Var17     float64
Var18     float64
Var19     float64
Var20     float64
Var21     float64
Var22     float64
Var23     float64
Var24     float64
Var25     float64
Var26     float64
Var27     float64
Var28     float64
Var29     float64
Var30     float64
           ...   
Var201     object
Var202     object
Var203     object
Var204     object
Var205     object
Var206     object
Var207     object
Var208     object
Var209    float64
Var210     object
Var211     object
Var212     object
Var213     object
Var214     object
Var215     object
Var216     object
Var217     object
Var218     object
Var219     object
Var220     object
Var221     object
Var222     object
Var223     object
Var224     object
Var225    

In [7]:
object_columns = train_data.select_dtypes(['object']).columns
object_columns

Index(['Var191', 'Var192', 'Var193', 'Var194', 'Var195', 'Var196', 'Var197',
       'Var198', 'Var199', 'Var200', 'Var201', 'Var202', 'Var203', 'Var204',
       'Var205', 'Var206', 'Var207', 'Var208', 'Var210', 'Var211', 'Var212',
       'Var213', 'Var214', 'Var215', 'Var216', 'Var217', 'Var218', 'Var219',
       'Var220', 'Var221', 'Var222', 'Var223', 'Var224', 'Var225', 'Var226',
       'Var227', 'Var228', 'Var229'],
      dtype='object')

In [11]:
train_data[object_columns] = train_data[object_columns].apply(lambda x: x.astype('category'))    
train_data

Var1  Var2  Var3  Var4  Var5     Var6  Var7  Var8  Var9  Var10  ...  \
0       NaN   NaN   NaN   NaN   NaN   1526.0   7.0   NaN   NaN    NaN  ...   
1       NaN   NaN   NaN   NaN   NaN    525.0   0.0   NaN   NaN    NaN  ...   
2       NaN   NaN   NaN   NaN   NaN   5236.0   7.0   NaN   NaN    NaN  ...   
3       NaN   NaN   NaN   NaN   NaN      NaN   0.0   NaN   NaN    NaN  ...   
4       NaN   NaN   NaN   NaN   NaN   1029.0   7.0   NaN   NaN    NaN  ...   
5       NaN   NaN   NaN   NaN   NaN    658.0   7.0   NaN   NaN    NaN  ...   
6       NaN   NaN   NaN   NaN   NaN   1680.0   7.0   NaN   NaN    NaN  ...   
7       NaN   NaN   NaN   NaN   NaN     77.0   0.0   NaN   NaN    NaN  ...   
8       NaN   NaN   NaN   NaN   NaN   1176.0   7.0   NaN   NaN    NaN  ...   
9       NaN   NaN   NaN   NaN   NaN   1141.0   7.0   NaN   NaN    NaN  ...   
10      NaN   NaN   NaN   NaN   NaN    490.0   7.0   NaN   NaN    NaN  ...   
11      NaN   NaN   NaN   NaN   NaN    798.0  14.0   NaN   NaN    NaN  ...   
12      NaN   NaN   NaN   NaN   NaN    595.0   0.0   NaN   NaN    NaN  ...   
13      NaN   NaN   NaN   NaN   NaN   2268.0   0.0   NaN   NaN    NaN  ...   
14      NaN   NaN   NaN   NaN   NaN   3633.0   7.0   NaN   NaN    NaN  ...   
15      NaN   NaN   NaN   NaN   NaN    259.0   0.0   NaN   NaN    NaN  ...   
16      NaN   NaN   NaN   NaN   NaN   5152.0   7.0   NaN   NaN    NaN  ...   
17      NaN   NaN   NaN   NaN   NaN   1449.0   7.0   NaN   NaN    NaN  ...   
18      NaN   NaN   NaN   NaN   NaN    574.0   7.0   NaN   NaN    NaN  ...   
19      NaN   NaN   NaN   NaN   NaN    658.0   7.0   NaN   NaN    NaN  ...   
20      NaN   NaN   NaN   NaN   NaN    245.0   0.0   NaN   NaN    NaN  ...   
21      NaN   NaN   NaN   NaN   NaN     14.0   0.0   NaN   NaN    NaN  ...   
22      NaN   NaN   NaN   NaN   NaN    826.0   0.0   NaN   NaN    NaN  ...   
23      NaN   NaN   NaN   NaN   NaN    854.0   7.0   NaN   NaN    NaN  ...   
24      NaN   NaN   NaN   NaN   NaN   1218.0   7.0   NaN   NaN    NaN  ...   
25      NaN   NaN   NaN   NaN   NaN   1043.0   7.0   NaN   NaN    NaN  ...   
26      NaN   NaN   NaN   0.0   NaN      NaN   NaN   NaN   NaN    NaN  ...   
27      NaN   NaN   NaN   NaN   NaN   1533.0   7.0   NaN   NaN    NaN  ...   
28      NaN   NaN   NaN   NaN   NaN    161.0   0.0   NaN   NaN    NaN  ...   
29      NaN   NaN   NaN   NaN   NaN     63.0   0.0   NaN   NaN    NaN  ...   
...     ...   ...   ...   ...   ...      ...   ...   ...   ...    ...  ...   
24970   NaN   NaN   NaN   NaN   NaN   1715.0   7.0   NaN   NaN    NaN  ...   
24971   NaN   NaN   NaN   NaN   NaN   2471.0  14.0   NaN   NaN    NaN  ...   
24972   NaN   NaN   NaN   NaN   NaN    784.0  21.0   NaN   NaN    NaN  ...   
24973   NaN   NaN   NaN   NaN   NaN    280.0   7.0   NaN   NaN    NaN  ...   
24974   NaN   NaN   NaN   NaN   NaN    826.0   7.0   NaN   NaN    NaN  ...   
24975   NaN   NaN   NaN   NaN   NaN      0.0   0.0   NaN   NaN    NaN  ...   
24976   NaN   NaN   NaN   NaN   NaN   1036.0   7.0   NaN   NaN    NaN  ...   
24977   NaN   NaN   NaN   NaN   NaN   1169.0   7.0   NaN   NaN    NaN  ...   
24978   NaN   NaN   NaN   NaN   NaN    805.0   7.0   NaN   NaN    NaN  ...   
24979   NaN   NaN   NaN   NaN   NaN    406.0   0.0   NaN   NaN    NaN  ...   
24980   NaN   NaN   NaN   NaN   0.0      NaN   NaN   NaN   NaN    0.0  ...   
24981   NaN   NaN   NaN   NaN   NaN    175.0   7.0   NaN   NaN    NaN  ...   
24982   NaN   NaN   NaN   NaN   NaN   1778.0   7.0   NaN   NaN    NaN  ...   
24983   NaN   NaN   NaN   NaN   NaN   1715.0   7.0   NaN   NaN    NaN  ...   
24984   NaN   NaN   NaN   NaN   NaN   1358.0   7.0   NaN   NaN    NaN  ...   
24985   NaN   NaN   NaN   NaN   NaN    518.0   0.0   NaN   NaN    NaN  ...   
24986   NaN   NaN   NaN   NaN   NaN    161.0   0.0   NaN   NaN    NaN  ...   
24987   NaN   NaN   NaN   NaN   NaN   1015.0  14.0   NaN   NaN    NaN  ...   
24988   NaN   NaN   NaN   NaN   NaN   1911.0   7.0   NaN   NaN    NaN  ...   
24989   NaN   NaN   NaN   NaN

In [12]:

train_data.dtypes

Var1       float64
Var2       float64
Var3       float64
Var4       float64
Var5       float64
Var6       float64
Var7       float64
Var8       float64
Var9       float64
Var10      float64
Var11      float64
Var12      float64
Var13      float64
Var14      float64
Var15      float64
Var16      float64
Var17      float64
Var18      float64
Var19      float64
Var20      float64
Var21      float64
Var22      float64
Var23      float64
Var24      float64
Var25      float64
Var26      float64
Var27      float64
Var28      float64
Var29      float64
Var30      float64
            ...   
Var201    category
Var202    category
Var203    category
Var204    category
Var205    category
Var206    category
Var207    category
Var208    category
Var209     float64
Var210    category
Var211    category
Var212    category
Var213    category
Var214    category
Var215    category
Var216    category
Var217    category
Var218    category
Var219    category
Var220    category
Var221    category
Var222    ca

In [1]:
train_data[object_columns] = train_data[object_columns].apply(lambda x: x.cat.codes)

NameError: name 'train_data' is not defined

In [14]:
train_data.dtypes

Var1      float64
Var2      float64
Var3      float64
Var4      float64
Var5      float64
Var6      float64
Var7      float64
Var8      float64
Var9      float64
Var10     float64
Var11     float64
Var12     float64
Var13     float64
Var14     float64
Var15     float64
Var16     float64
Var17     float64
Var18     float64
Var19     float64
Var20     float64
Var21     float64
Var22     float64
Var23     float64
Var24     float64
Var25     float64
Var26     float64
Var27     float64
Var28     float64
Var29     float64
Var30     float64
           ...   
Var201       int8
Var202      int16
Var203       int8
Var204       int8
Var205       int8
Var206       int8
Var207       int8
Var208       int8
Var209    float64
Var210       int8
Var211       int8
Var212       int8
Var213       int8
Var214      int16
Var215       int8
Var216      int16
Var217      int16
Var218       int8
Var219       int8
Var220      int16
Var221       int8
Var222      int16
Var223       int8
Var224       int8
Var225    

In [ ]:
string_columns = train_data.loc[:,'Var191':'Var230']


In [ ]:
x = train_data.loc[:,'Var228'].astype('category')

In [ ]:
x = x.astype('category').cat.codes
train_data.loc[:, 'Var230'] = x
train_data.loc[:, 'Var230'].describe()

In [ ]:
x.head(100)

#### Load testing data

In [ ]:
test_data = pd.read_csv('orange_small_test.data', sep = '\t')
test_data

#### Load training lables

In [ ]:
appetency = pd.read_csv('orange_small_train_appetency_train.labels', sep = '\t').astype('float')

In [ ]:
appetency

In [ ]:
#appetency = (appetency + 1)/2

In [ ]:
appetency.describe()

In [ ]:
lgb_train_data = lgb.Dataset(data=train_data, label=appetency)
lgb_test_data = lgb.Dataset(data=test_data)

In [ ]:
param = {'num_leaves': 31, 'objective': 'binary','metric':'auc'}

In [ ]:
num_round = 10
bst = lgb.train(param, lgb_train_data, num_round, valid_sets=[lgb_test_data])

In [ ]:
churn = pd.read_table('orange_small_train_churn_train.labels',header = None).loc[:,0].astype('category')

In [ ]:
churn.describe()

In [ ]:
upselling = pd.read_table('orange_small_train_upselling_train.labels',header = None).loc[:,0].astype('category')

In [ ]:
upselling.describe()